src/Features/Core/Portable/CodeRefactorings/AddMissingImports/AbstractAddMissingImportsFeatureService.cs

```diff
@@ -0,0 +1,231 @@
+                newSolution = newProject.Solution;
+            }
+
+            return newSolution;
+        }
+
+        public async Task<Project> AddMissingImportsAsync(Project project, CancellationToken cancellationToken)
+        {
+            var newProject = project;
+
+            foreach (var documentId in project.DocumentIds)
+            {
+                var document = newProject.GetDocument(documentId);
+                newProject = await AddMissingImportsAsync(document, cancellationToken).ConfigureAwait(false);
+            }
+
+            return newProject;
+        }
+
+        public async Task<Project> AddMissingImportsAsync(Document document, CancellationToken cancellationToken)
+        {
+            var diagnostics = await GetDiagnosticsAsync(document, cancellationToken).ConfigureAwait(false);
+            return await FixDocument(document, diagnostics, cancellationToken).ConfigureAwait(false);
+        }
+
+        public async Task<Project> AddMissingImportsAsync(Document document, TextSpan textSpan, CancellationToken cancellationToken)
+        {
+            var diagnostics = await GetDiagnosticsAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
+            return await FixDocument(document, diagnostics, cancellationToken).ConfigureAwait(false);
+        }
+
+        public async Task<bool> IsMissingImportsAsync(Document document, CancellationToken cancellationToken)
+        {
+            var diagnostics = await GetDiagnosticsAsync(document, cancellationToken).ConfigureAwait(false);
+            return !diagnostics.IsEmpty;
+        }
+
+        public async Task<bool> IsMissingImportsAsync(Document document, TextSpan textSpan, CancellationToken cancellationToken)
+        {
+            var diagnostics = await GetDiagnosticsAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
+            return !diagnostics.IsEmpty;
+        }
+
+        private Task<ImmutableArray<Diagnostic>> GetDiagnosticsAsync(Document document, CancellationToken cancellationToken)
+        {
+            if (!document.TryGetText(out var text))
+            {
+                return Task.FromResult(ImmutableArray<Diagnostic>.Empty);
+            }
+
+            var textSpan = new TextSpan(0, text.Length);
+            return GetDiagnosticsAsync(document, textSpan, cancellationToken);
+        }
+
+        private async Task<ImmutableArray<Diagnostic>> GetDiagnosticsAsync(Document document, TextSpan textSpan, CancellationToken cancellationToken)
+        {
+            if (!document.SupportsSemanticModel)
+            {
+                return ImmutableArray<Diagnostic>.Empty;
+            }
+
+            var project = document.Project;
+
+            var diagnosticData = await _diagnosticAnalyzerService.GetDiagnosticsForSpanAsync(document, textSpan, cancellationToken: cancellationToken).ConfigureAwait(false);
```

Comment:
is this new code or existing code? just your FYI. nested concurrency is usually worse than no concurrency. make sure when you do concurrency, there is no other one trying concurrency up in your stack.